core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -148,9 +161,23 @@ private[spark] class Executor(
   }
 
   def killTask(taskId: Long, interruptThread: Boolean): Unit = {
-    val tr = runningTasks.get(taskId)
-    if (tr != null) {
-      tr.kill(interruptThread)
+    val taskRunner = runningTasks.get(taskId)
+    if (taskRunner != null) {
+      if (taskReaperEnabled) {
+        taskReaperForTask.synchronized {
+          val shouldCreateReaper = taskReaperForTask.get(taskId) match {
+            case None => true
+            case Some(existingReaper) => interruptThread && !existingReaper.interruptThread
+          }
+          if (shouldCreateReaper) {
+            val taskReaper = new TaskReaper(taskRunner, interruptThread = interruptThread)
+            taskReaperForTask(taskId) = taskReaper
+            taskReaperPool.execute(taskReaper)
+          }
+        }
+      } else {
+        taskRunner.kill(interruptThread = interruptThread)
```

Comment:
Do we want to introduce dumping stack for this path ? I really like the idea of doing that !